File: csharp/deserialize.cs
/*
	Deserialize - load bytecode produced by compiler.

	Copyright (c) 2022 Frank McIngvale, see LICENSE

	Ported from C++ implementation.
*/
#nullable enable

using System;
using System.IO;

public class Deserializer {

	public static bool parseBool(string text) {
		if(text == "true")
			return true;
		else if(text == "false")
			return false;
		else
			throw new LangError("Bad boolean literal: " + text);
	}

	// deserialize & return next object from stream
	public static LangObject deserialize_stream(Interpreter intr, StreamReader stream) {
		string? line;
		while(true) {
			line = stream.ReadLine();
			if(line == null) {
				return new LangVoid();
			}
			// remove any \n or \r left on line
			while(line.Length > 0 && (line[line.Length-1] == '\n' || line[line.Length-1] == '\r'))
				line.Remove(line.Length-1);
			
			switch(line[0]) {
				case 'M':
					// ignore metadata and return NEXT object
					return deserialize_stream(intr, stream);
				case 'i': return new LangInt(long.Parse(line.Substring(2)));
				case 'f': return new LangFloat(double.Parse(line.Substring(2)));
				case 'b': return new LangBool(parseBool(line.Substring(2)));
				case 'n': return new LangNull();
				case 'o': 
					{ 
						byte code, A;
						ushort B;
						uint C;
						Opcodes.opcode_unpack(long.Parse(line.Substring(2)),
											out code, out A, out B, out C);
						return new LangOpcode(code, A, B, C);
					}
				case 's':
					line = line.Substring(2);
					line = line.Replace("%32", " ");
					line = line.Replace("%09", "\t");
					line = line.Replace("%10", "\n");
					line = line.Replace("%13", "\r");
					line = line.Replace("%37", "%");
					return new LangString(line);
				case 'y': return new LangSymbol(line.Substring(2));
				case 'L':
					{
						int nr = int.Parse(line.Substring(2));
						var list = new LangList();
						for(int i=0; i<nr; ++i) {
							list.objlist.Add(deserialize_stream(intr, stream));
						}
						return list;
					}
				case 'F':
					{
						LangObject list = deserialize_stream(intr, stream);
						if(list is LangList) {
							return new LangLambda((list as LangList)!.objlist);
						}
						else {
							throw new LangError("Expecting list after F but got: " + list.fmtStackPrint());
						}
					}
				case 'W':
					{
						string name = line.Substring(2);
						LangObject list = deserialize_stream(intr, stream);
						if(list is LangList) {
							// do not allow overwriting words
							intr.defineWord(name, (list as LangList)!.objlist, false);
							return new LangVoid();
						}
						else {
							throw new LangError("Expecting list after W but got: " + list.fmtStackPrint());
						}
					}
				default:
					throw new LangError("Unrecognized line in deserialize: " + line);
			}
		}
	}
}

File: python/deserialize.py
from __future__ import annotations
from errors import LangError
from langtypes import LangLambda, LangString, isList, parseBool, LangNull, LangVoid, \
				LangOpcode
"""
	Deserialize - load bytecode from compiler and put into Interpreter.
	
	Copyright (c) 2022 Frank McIngvale, see LICENSE

	Ported from the C++ implementation.
"""

def deserialize_stream(intr, fileIn):
	line = fileIn.readline()
	if line == "":
		return LangVoid()

	line = line.rstrip() # remove \n
	if line[0] == 'i': return int(line[2:])
	elif line[0] == 'f': return float(line[2:])
	elif line[0] == 'b': return parseBool(line[2:])
	elif line[0] == 'n': return LangNull() 
	elif line[0] == 'o': return LangOpcode(int(line[2:]))
	elif line[0] == 's':
		s = line[2:]
		s = s.replace("%32"," ").replace("%09","\t").replace("%10","\n").replace("%13","\r").replace("%37","%")
		return LangString(s)
	elif line[0] == 'y': return line[2:]
	elif line[0] == 'L':
		nr = int(line[2:])
		objs = []
		for i in range(nr):
			objs.append(deserialize_stream(intr, fileIn))
		return objs
	elif line[0] == 'F': # lambda
		objs = deserialize_stream(intr, fileIn)
		if not isList(objs):
			raise LangError("Expecting list after 'F' but got: " + objs.fmtStackPrint())
		return LangLambda(objs)
	elif line[0] == 'W': # word
		name = line[2:]
		objs = deserialize_stream(intr, fileIn)
		if not isList(objs):
			raise LangError("Expecting list after 'W' but got: " + objs.fmtStackPrint())
		intr.defineWord(name,objs,False)
		#print("LOADED WORD: " + name)
		return LangVoid()
	else:
		raise LangError("Unrecognized line while deserializing: " + line)


